File: DSA/Assignment_5/doubly_rev.c
#include <stdio.h>
#include <stdlib.h>
typedef struct Node
{
    int data;
    struct Node *prev;
    struct Node *next;
}Node;

Node *createNode(int data)
{
    Node *newNode=(Node *)malloc(sizeof(Node));
    newNode->data=data;
    newNode->prev=NULL;
    newNode->next=NULL;
    return newNode;
}

void insert_ele(Node* *head, int data)
{
    Node *newNode=createNode(data);
    if(*head==NULL)
    {
        *head=newNode;
        return;
    }
    Node *temp=*head;
    while(temp->next!=NULL)
    temp=temp->next;
    temp->next=newNode;
}

void print_list(Node* *head)
{
    Node *temp=*head;
    while(temp)
    {
        printf("%d ",temp->data);
        temp=temp->next;
    }
    printf("\n");
}

void rev(Node* *head)
{
    if(*head==NULL) return;
    Node *temp=*head;
    Node *last=NULL;
    while(temp)
    {
        Node *swap=temp->next;
        temp->next=temp->prev;
        temp->prev=swap;

        last=temp;
        temp=temp->prev;
    }
    *head=last;
}

int main()
{
    int n, data;
    
    
    return 0;
}

File: DSA/Assignment_5/doubly_rev.py
class Node:
    def __init__(self, data):
        self.data=data
        self.next=None
        self.prev=None

class DoublyLinkedList():
    def __init__(self):
        self.head=None

    def createNode(self, data):
        return Node(data)
    
    def insert_ele(self, data):
        newNode=self.createNode(data)
        if self.head is None:
            self.head=newNode
            return
        temp=self.head
        while temp.next:
            temp=temp.next
        temp.next=newNode
        newNode.prev=temp

    def print_list(self):
        if self.head is None:
            print("Empty List!")
            return
        temp=self.head
        while temp:
            print(temp.data,end=' ')
            temp=temp.next
        print()

    def rev(self):
        if self.heada is None:
            print("Empty List!")
            return
        temp=self.head
        last=None
        while temp:
            swap=temp.next
            temp.next=temp.prev
            temp.prev=swap

            last=temp
            temp=temp.prev
        last=self.head

def main():
    list=DoublyLinkedList()

    #Formation of doubly linked list
    n=int(input("Enter number of nodes:"))
    for _ in range(n):
        ele=int(input("Enter element:"))
        list.insert_ele(ele)
    print("Original List:",end='')
    list.print_list()

    #Reverse doubly linked list
    list.rev()

    #Print reversed doubly linked list
    print("Reversed List:",end=)